File: backend/src/controllers/salasController.js
import { pool } from '../config/database.js';

// Obtener todas las salas (con filtro opcional por tipo)
export const obtenerSalas = async (req, res) => {
  try {
    const { tipo_sala } = req.query;

    let query = `
      SELECT
        s.nombre_sala,
        s.edificio,
        s.capacidad,
        s.tipo_sala,
        e.nombre_edificio,
        e.direccion
      FROM sala s
      INNER JOIN edificio e ON s.edificio = e.nombre_edificio
    `;

    const params = [];

    if (tipo_sala) {
      query += ' WHERE s.tipo_sala = ?';
      params.push(tipo_sala);
    }

    query += ' ORDER BY s.edificio, s.nombre_sala';

    const [salas] = await pool.query(query, params);

    res.json({
      success: true,
      salas
    });

  } catch (error) {
    console.error('Error al obtener salas:', error);
    res.status(500).json({
      error: 'Error en el servidor'
    });
  }
};

// Obtener disponibilidad de una sala específica en una fecha
export const obtenerDisponibilidadSala = async (req, res) => {
  try {
    const { nombre } = req.params;
    const { fecha, edificio } = req.query;

    if (!fecha || !edificio) {
      return res.status(400).json({
        error: 'Fecha y edificio son requeridos'
      });
    }

    // Verificar que la sala exista
    const [salaRows] = await pool.query(
      'SELECT * FROM sala WHERE nombre_sala = ? AND edificio = ?',
      [nombre, edificio]
    );

    if (salaRows.length === 0) {
      return res.status(404).json({
        error: 'Sala no encontrada'
      });
    }

    // Obtener todos los turnos
    const [turnos] = await pool.query(
      'SELECT id_turno, hora_inicio, hora_fin FROM turno ORDER BY hora_inicio'
    );

    // Obtener reservas existentes para esa sala y fecha
    const [reservas] = await pool.query(
      `SELECT id_turno FROM reserva
       WHERE nombre_sala = ? AND edificio = ? AND fecha = ?
       AND estado IN ('activa', 'finalizada')`,
      [nombre, edificio, fecha]
    );

    const turnosReservados = new Set(reservas.map(r => r.id_turno));

    // Marcar turnos como disponibles o no
    const turnosConDisponibilidad = turnos.map(turno => ({
      id_turno: turno.id_turno,
      hora_inicio: turno.hora_inicio,
      hora_fin: turno.hora_fin,
      disponible: !turnosReservados.has(turno.id_turno)
    }));

    res.json({
      success: true,
      sala: nombre,
      edificio,
      fecha,
      turnos: turnosConDisponibilidad
    });

  } catch (error) {
    console.error('Error al obtener disponibilidad:', error);
    res.status(500).json({
      error: 'Error en el servidor'
    });
  }
};

// Obtener todos los edificios
export const obtenerEdificios = async (req, res) => {
  try {
    const [edificios] = await pool.query(
      'SELECT nombre_edificio, direccion FROM edificio ORDER BY nombre_edificio'
    );

    res.json({
      success: true,
      edificios
    });

  } catch (error) {
    console.error('Error al obtener edificios:', error);
    res.status(500).json({
      error: 'Error en el servidor'
    });
  }
};

// Obtener salas de un edificio específico
export const obtenerSalasPorEdificio = async (req, res) => {
  try {
    const { edificio } = req.params;

    const [salas] = await pool.query(
      `SELECT
        nombre_sala,
        edificio,
        capacidad,
        tipo_sala
      FROM sala
      WHERE edificio = ?
      ORDER BY nombre_sala`,
      [edificio]
    );

    res.json({
      success: true,
      edificio,
      salas
    });

  } catch (error) {
    console.error('Error al obtener salas del edificio:', error);
    res.status(500).json({
      error: 'Error en el servidor'
    });
  }
};

File: backendP/src/controllers/salas_controller.py
from fastapi import HTTPException, Query
from src.config.database import get_connection

async def obtener_salas(tipo_sala: str = None):
    """Obtener todas las salas (con filtro opcional por tipo)"""
    try:
        conn = get_connection()
        cursor = conn.cursor(dictionary=True)

        query = """
            SELECT
                s.nombre_sala,
                s.edificio,
                s.capacidad,
                s.tipo_sala,
                e.nombre_edificio,
                e.direccion
            FROM sala s
            INNER JOIN edificio e ON s.edificio = e.nombre_edificio
        """

        params = []

        if tipo_sala:
            query += ' WHERE s.tipo_sala = %s'
            params.append(tipo_sala)

        query += ' ORDER BY s.edificio, s.nombre_sala'

        cursor.execute(query, params)
        salas = cursor.fetchall()

        cursor.close()
        conn.close()

        return {
            "success": True,
            "salas": salas
        }

    except Exception as error:
        print(f'Error al obtener salas: {error}')
        raise HTTPException(
            status_code=500,
            detail="Error en el servidor"
        )


async def obtener_disponibilidad_sala(nombre: str, fecha: str, edificio: str):
    """Obtener disponibilidad de una sala específica en una fecha"""
    try:
        if not fecha or not edificio:
            raise HTTPException(
                status_code=400,
                detail="Fecha y edificio son requeridos"
            )

        conn = get_connection()
        cursor = conn.cursor(dictionary=True)

        # Verificar que la sala exista
        cursor.execute(
            'SELECT * FROM sala WHERE nombre_sala = %s AND edificio = %s',
            (nombre, edificio)
        )
        sala_rows = cursor.fetchall()

        if len(sala_rows) == 0:
            cursor.close()
            conn.close()
            raise HTTPException(
                status_code=404,
                detail="Sala no encontrada"
            )

        # Obtener todos los turnos
        cursor.execute(
            'SELECT id_turno, hora_inicio, hora_fin FROM turno ORDER BY hora_inicio'
        )
        turnos = cursor.fetchall()

        # Obtener reservas existentes para esa sala y fecha
        cursor.execute(
            """SELECT id_turno FROM reserva
            WHERE nombre_sala = %s AND edificio = %s AND fecha = %s
            AND estado IN ('activa', 'finalizada')""",
            (nombre, edificio, fecha)
        )
        reservas = cursor.fetchall()

        turnos_reservados = {r['id_turno'] for r in reservas}

        # Marcar turnos como disponibles o no
        turnos_con_disponibilidad = [
            {
                "id_turno": turno['id_turno'],
                "hora_inicio": str(turno['hora_inicio']),
                "hora_fin": str(turno['hora_fin']),
                "disponible": turno['id_turno'] not in turnos_reservados
            }
            for turno in turnos
        ]

        cursor.close()
        conn.close()

        return {
            "success": True,
            "sala": nombre,
            "edificio": edificio,
            "fecha": fecha,
            "turnos": turnos_con_disponibilidad
        }

    except HTTPException:
        raise
    except Exception as error:
        print(f'Error al obtener disponibilidad: {error}')
        raise HTTPException(
            status_code=500,
            detail="Error en el servidor"
        )


async def obtener_edificios():
    """Obtener todos los edificios"""
    try:
        conn = get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            'SELECT nombre_edificio, direccion FROM edificio ORDER BY nombre_edificio'
        )
        edificios = cursor.fetchall()

        cursor.close()
        conn.close()

        return {
            "success": True,
            "edificios": edificios
        }

    except Exception as error:
        print(f'Error al obtener edificios: {error}')
        raise HTTPException(
            status_code=500,
            detail="Error en el servidor"
        )


async def obtener_salas_por_edificio(edificio: str):
    """Obtener salas de un edificio específico"""
    try:
        conn = get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            """SELECT
                nombre_sala,
                edificio,
                capacidad,
                tipo_sala
            FROM sala
            WHERE edificio = %s
            ORDER BY nombre_sala""",
            (edificio,)
        )
        salas = cursor.fetchall()

        cursor.close()
        conn.close()

        return {
            "success": True,
            "edificio": edificio,
            "salas": salas
        }

    except Exception as error:
        print(f'Error al obtener salas del edificio: {error}')
        raise HTTPException(
            status_code=500,
            detail="Error en el servidor"
        )

